File: src/com/ptby/dynamicreturntypeplugin/json/FileSytemConfigChangeListener.java
package com.ptby.dynamicreturntypeplugin.json;

import com.intellij.openapi.vfs.*;

import java.util.ArrayList;
import java.util.List;

public class FileSytemConfigChangeListener implements VirtualFileListener {
    private static final String expectedConfigFileName = "dynamicReturnTypeMeta.json";
    private List<ProjectConfigChangeListener> projectConfigChangeListeners = new ArrayList<ProjectConfigChangeListener>();

    public FileSytemConfigChangeListener(){
        LocalFileSystem instance = LocalFileSystem.getInstance();
        instance.addVirtualFileListener( this );
    }


    @Override
    public void propertyChanged( VirtualFilePropertyEvent event ) {
    }

    public void registerProjectConfigChangeListener( ProjectConfigChangeListener projectConfigChangeListener ) {
        projectConfigChangeListeners.add( projectConfigChangeListener );
    }


    @Override
    public void contentsChanged( VirtualFileEvent event ) {
        refreshIfCorrectFileEvent( event );
    }


    private void refreshIfCorrectFileEvent( VirtualFileEvent virtualFileEvent ) {
        String currentFileName = virtualFileEvent.getFileName();

        if ( currentFileName.equals( expectedConfigFileName ) ) {
            refreshConfigs();
        }
    }


    private void refreshConfigs() {
        for ( ProjectConfigChangeListener projectConfigChangeListener : projectConfigChangeListeners ) {
            projectConfigChangeListener.refreshAllConfigs();
        }
    }


    @Override
    public void fileCreated( VirtualFileEvent event ) {
        refreshIfCorrectFileEvent( event );
    }

    @Override
    public void fileDeleted( VirtualFileEvent event ) {
        refreshIfCorrectFileEvent( event );
    }

    @Override
    public void fileMoved( VirtualFileMoveEvent event ) {
        refreshIfCorrectFileEvent( event );
    }

    @Override
    public void fileCopied( VirtualFileCopyEvent event ) {
        refreshIfCorrectFileEvent( event );
    }

    @Override
    public void beforePropertyChange( VirtualFilePropertyEvent event ) {

    }

    @Override
    public void beforeContentsChange( VirtualFileEvent event ) {

    }

    @Override
    public void beforeFileDeletion( VirtualFileEvent event ) {

    }

    @Override
    public void beforeFileMovement( VirtualFileMoveEvent event ) {

    }
}

File: src/com/ptby/dynamicreturntypeplugin/json/FileSytemConfigChangeListener.kt
package com.ptby.dynamicreturntypeplugin.json

import com.intellij.openapi.vfs.*
import java.util.*

public class FileSytemConfigChangeListener : VirtualFileListener {
    private val projectConfigChangeListeners = ArrayList<ProjectConfigChangeListener>()

    init {
        val instance = LocalFileSystem.getInstance()
        instance.addVirtualFileListener(this)
    }


    override fun propertyChanged(event: VirtualFilePropertyEvent) {
    }

    public fun registerProjectConfigChangeListener(projectConfigChangeListener: ProjectConfigChangeListener) {
        projectConfigChangeListeners.add(projectConfigChangeListener)
    }


    override fun contentsChanged(event: VirtualFileEvent) {
        refreshIfCorrectFileEvent(event)
    }


    private fun refreshIfCorrectFileEvent(virtualFileEvent: VirtualFileEvent) {
        val currentFileName = virtualFileEvent.getFileName()

        if (currentFileName == expectedConfigFileName) {
            refreshConfigs()
        }
    }


    private fun refreshConfigs() {
        for (projectConfigChangeListener in projectConfigChangeListeners) {
            projectConfigChangeListener.refreshAllConfigs()
        }
    }


    override fun fileCreated(event: VirtualFileEvent) {
        refreshIfCorrectFileEvent(event)
    }

    override fun fileDeleted(event: VirtualFileEvent) {
        refreshIfCorrectFileEvent(event)
    }

    override fun fileMoved(event: VirtualFileMoveEvent) {
        refreshIfCorrectFileEvent(event)
    }

    override fun fileCopied(event: VirtualFileCopyEvent) {
        refreshIfCorrectFileEvent(event)
    }

    override fun beforePropertyChange(event: VirtualFilePropertyEvent) {
    }

    override fun beforeContentsChange(event: VirtualFileEvent) {
    }

    override fun beforeFileDeletion(event: VirtualFileEvent) {
    }

    override fun beforeFileMovement(event: VirtualFileMoveEvent) {
    }

    companion object {
        private val expectedConfigFileName = "dynamicReturnTypeMeta.json"
    }
}
